master/internal/api_project.go

```diff
@@ -535,9 +536,26 @@ func (a *apiServer) PostProject(
 		return nil, status.Error(codes.PermissionDenied, err.Error())
 	}
 
+	var projectKey string
+	if req.Key == nil {
+		projectKey, err = project.GenerateProjectKey(ctx, req.Name)
+		if err != nil {
+			return nil, fmt.Errorf("error generating project key: %w", err)
+		}
+	} else {
+		projectKey = *req.Key
+	}
+
 	p := &projectv1.Project{}
 	err = a.m.db.QueryProto("insert_project", p, req.Name, req.Description,
-		req.WorkspaceId, curUser.ID)
+		req.WorkspaceId, curUser.ID, projectKey)
```

Comment:
Worried about a race condition here where two projects with the same first 3 chars are being created at the same time. Both of them would generate the same key. Perhaps we can create the project first without the key, then generates the key filtering out projects created after its own "created_at" time.

Reply:
I agree that the race condition here will cause a rough UX, "if no key is specified during project creation, but getting an error about a key conflict" doesn't really make sense.

The suggest approach would require dropping the unique constraint on the key column (which might not be the worst move, since I think that is really just a bandaid on an application level solution for how we handle uniqueness on project keys). I'd also lean away from using the created_at time, since it's susceptible to clock drift/syncs, and could cause some funky flakes. We could piggy back of the strict event ordering work done by streaming updates and leverage the seq value (would do the same thing we'd want to do with a timestamp, without the risk of clock drift).